src/process_collector.rs

```diff
@@ -0,0 +1,358 @@
+/// the process start time for the given process id.
+pub struct ProcessCollector {
+    pid: pid_t,
+    descs: Vec<Desc>,
+    cpu_total: Mutex<Counter>,
+    open_fds: Gauge,
+    max_fds: Gauge,
+    vsize: Gauge,
+    rss: Gauge,
+    start_time: Gauge,
+}
+
+impl ProcessCollector {
+    /// Create a `ProcessCollector` with the given process id and namespace.
+    pub fn new<S: Into<String>>(pid: pid_t, namespace: S) -> ProcessCollector {
+        let namespace = namespace.into();
+        let mut descs = Vec::new();
+
+        let cpu_total = Counter::with_opts(Opts::new("process_cpu_seconds_total",
+                                                     "Total user and system CPU time spent in \
+                                                      seconds.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(cpu_total.desc().into_iter().cloned());
+
+        let open_fds = Gauge::with_opts(Opts::new("process_open_fds",
+                                                  "Number of open file descriptors.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(open_fds.desc().into_iter().cloned());
+
+        let max_fds = Gauge::with_opts(Opts::new("process_max_fds",
+                                                 "Maximum number of open file descriptors.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(max_fds.desc().into_iter().cloned());
+
+        let vsize = Gauge::with_opts(Opts::new("process_virtual_memory_bytes",
+                                               "Virtual memory size in bytes.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(vsize.desc().into_iter().cloned());
+
+        let rss = Gauge::with_opts(Opts::new("process_resident_memory_bytes",
+                                             "Resident memory size in bytes.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(rss.desc().into_iter().cloned());
+
+        let start_time = Gauge::with_opts(Opts::new("process_start_time_seconds",
+                                                    "Start time of the process since unix epoch \
+                                                     in seconds.")
+                .namespace(namespace.clone()))
+            .unwrap();
+        descs.extend(start_time.desc().into_iter().cloned());
+
+        ProcessCollector {
+            pid: pid,
+            descs: descs,
+            cpu_total: Mutex::new(cpu_total),
+            open_fds: open_fds,
+            max_fds: max_fds,
+            vsize: vsize,
+            rss: rss,
+            start_time: start_time,
+        }
+    }
+}
+
+impl Collector for ProcessCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
```

Comment:
implementation note: golang client uses constant metrics instead of modifying guages. Source: https://github.com/prometheus/client_golang/blob/master/prometheus/process_collector.go#L113-L140